File: retrocpu_test_framework/src/mn1613/m_sequence.ts
/**
 * テスト用メモリ初期化の 16bit M系列（Galois LFSR）。
 * bios_common.asm の g_get_rnd と同じ多項式 x^16+x^14+x^13+x^11+1（タップ 0xB400）。
 * 種は現在時刻から作る（0 はロックするので 1 にする）。
 */

/** Galois LFSR タップ（右シフト） */
export const MEM_MSEQ_TAP = 0xb400;

/**
 * 現在時刻から 16bit の M系列種を作る。0 にはしない。
 * @returns 1〜0xFFFF
 */
export function memMseqSeedFromTime(): number {
  const ns = process.hrtime.bigint();
  const ms = BigInt(Date.now());
  const mixed = Number((ns ^ (ms << 20n) ^ (ms >> 4n)) & 0xffffn);
  return mixed === 0 ? 1 : mixed;
}

/**
 * bios_common.asm と同じ右シフト Galois LFSR を 1 歩進める。
 * @param seed 種（16bit。0 は 1 にする）
 * @returns 次の値（1〜0xFFFF）
 */
export function mseqStep(seed: number): number {
  let x = seed & 0xffff;
  if (x === 0) {
    x = 1;
  }
  const lsb = x & 1;
  x >>>= 1;
  if (lsb !== 0) {
    x ^= MEM_MSEQ_TAP;
  }
  return x & 0xffff;
}

/**
 * 種を 16bit にし、0 なら 1 にする。
 * @param seed 任意の種
 * @returns 1〜0xFFFF
 */
function normalizeSeed(seed: number): number {
  const x = seed & 0xffff;
  return x === 0 ? 1 : x;
}

/**
 * バッファを 16bit ビッグエンディアンの M系列で埋める。
 * 物理ワード i には種から i+1 歩進めた値を書く。
 * @param buf 偶数バイト長のメモリ
 * @param seed 開始種。省略時は現在時刻から作る。0 は 1 にする
 * @returns 実際に使った種（1〜0xFFFF）
 */
export function fillMemoryMSequence(buf: ArrayBuffer, seed?: number): number {
  const start = normalizeSeed(seed ?? memMseqSeedFromTime());
  const bytes = new Uint8Array(buf);
  const words = buf.byteLength >>> 1;
  let x = start;
  let off = 0;
  for (let i = 0; i < words; i += 1) {
    x = mseqStep(x);
    bytes[off] = (x >>> 8) & 0xff;
    bytes[off + 1] = x & 0xff;
    off += 2;
  }
  return start;
}

/**
 * M系列で埋めたメモリを返す。種は省略時に現在時刻から決める。
 * @param byteLength バイト数（既定セッションは 256K ワード＝512KB）
 * @param seed 開始種。省略時は現在時刻
 * @returns buffer と実際の種
 */
export function createMSequenceMemory(
  byteLength: number,
  seed?: number,
): { buffer: ArrayBuffer; seed: number } {
  const buffer = new ArrayBuffer(byteLength);
  const used = fillMemoryMSequence(buffer, seed);
  return { buffer, seed: used };
}

File: retrocpu_test_framework_rs/src/mn1613/m_sequence.rs
use std::time::{SystemTime, UNIX_EPOCH};

use crate::error::FrameworkError;

pub const MEM_MSEQ_TAP: u16 = 0xb400;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MSequenceMemory {
    pub buffer: Vec<u8>,
    pub seed: u16,
}

pub fn mem_mseq_seed_from_time() -> u16 {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default();
    let mixed = (now.as_nanos() ^ ((now.as_millis()) << 20) ^ (now.as_millis() >> 4)) & 0xffff;
    let seed = mixed as u16;
    if seed == 0 {
        1
    } else {
        seed
    }
}

pub fn mseq_step(seed: u16) -> u16 {
    let mut x = if seed == 0 { 1 } else { seed };
    let lsb = x & 1;
    x >>= 1;
    if lsb != 0 {
        x ^= MEM_MSEQ_TAP;
    }
    x
}

fn normalize_seed(seed: u16) -> u16 {
    if seed == 0 {
        1
    } else {
        seed
    }
}

pub fn fill_memory_m_sequence(buf: &mut [u8], seed: Option<u16>) -> Result<u16, FrameworkError> {
    if (buf.len() & 1) != 0 {
        return Err(FrameworkError::invalid_argument(format!(
            "buffer length must be even (got {})",
            buf.len()
        )));
    }

    let start = normalize_seed(seed.unwrap_or_else(mem_mseq_seed_from_time));
    let mut x = start;

    for word_idx in 0..(buf.len() / 2) {
        x = mseq_step(x);
        let off = word_idx * 2;
        buf[off] = (x >> 8) as u8;
        buf[off + 1] = (x & 0xff) as u8;
    }

    Ok(start)
}

pub fn create_m_sequence_memory(
    byte_length: usize,
    seed: Option<u16>,
) -> Result<MSequenceMemory, FrameworkError> {
    if (byte_length & 1) != 0 {
        return Err(FrameworkError::invalid_argument(format!(
            "byte_length must be even (got {})",
            byte_length
        )));
    }

    let mut buffer = vec![0_u8; byte_length];
    let used_seed = fill_memory_m_sequence(&mut buffer, seed)?;
    Ok(MSequenceMemory {
        buffer,
        seed: used_seed,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn step_matches_known_values() {
        let mut x = 1_u16;
        x = mseq_step(x);
        assert_eq!(x, 0xb400);
        x = mseq_step(x);
        assert_eq!(x, 0x5a00);
    }

    #[test]
    fn fill_writes_big_endian_words() {
        let mut buf = vec![0_u8; 4];
        let seed = fill_memory_m_sequence(&mut buf, Some(1)).expect("fill should work");
        assert_eq!(seed, 1);
        assert_eq!(buf, vec![0xb4, 0x00, 0x5a, 0x00]);
    }

    #[test]
    fn odd_length_is_error() {
        let mut buf = vec![0_u8; 3];
        let err = fill_memory_m_sequence(&mut buf, Some(1)).unwrap_err();
        assert!(matches!(err, FrameworkError::InvalidArgument(_)));
    }
}
